File: eclipse/plugin/src/test/java/com/adityakumar/engine/WordDiffTest.java
package com.adityakumar.engine;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

/**
 * Unit tests for {@link WordDiff} — the inline word-level diff algorithm.
 *
 * <p>Mirrors the semantics documented in the VS Code DiffList.tsx
 * {@code computeInline()} specification.
 */
class WordDiffTest {

    // ------------------------------------------------------------------ //
    //  tokenize()                                                          //
    // ------------------------------------------------------------------ //

    @Test
    void tokenizeEmptyStringReturnsEmpty() {
        assertArrayEquals(new String[0], WordDiff.tokenize(""));
    }

    @Test
    void tokenizeWords() {
        String[] tokens = WordDiff.tokenize("hello world");
        assertArrayEquals(new String[]{"hello", " ", "world"}, tokens);
    }

    @Test
    void tokenizeMixedContent() {
        String[] tokens = WordDiff.tokenize("id=1000 name=alice");
        // Tokens: "id", "=", "1000", " ", "name", "=", "alice"
        assertArrayEquals(new String[]{"id", "=", "1000", " ", "name", "=", "alice"}, tokens);
    }

    @Test
    void tokenizePunctuation() {
        String[] tokens = WordDiff.tokenize("foo,bar;baz");
        assertArrayEquals(new String[]{"foo", ",", "bar", ";", "baz"}, tokens);
    }

    @Test
    void tokenizeWhitespaceRun() {
        String[] tokens = WordDiff.tokenize("a   b");
        assertArrayEquals(new String[]{"a", "   ", "b"}, tokens);
    }

    @Test
    void tokenizeReconstitutesOriginal() {
        String original = "  hello, world! 123_test  ";
        String[] tokens = WordDiff.tokenize(original);
        StringBuilder sb = new StringBuilder();
        for (String t : tokens) sb.append(t);
        assertEquals(original, sb.toString(),
                "tokens must concatenate back to the original string without gaps or overlaps");
    }

    // ------------------------------------------------------------------ //
    //  changedRange()                                                      //
    // ------------------------------------------------------------------ //

    @Test
    void changedRangeFullSpan() {
        String[] tokens = new String[]{"hello", " ", "world"};
        int[] span = WordDiff.changedRange(tokens, 0, 3);
        assertArrayEquals(new int[]{0, 11}, span);
    }

    @Test
    void changedRangeMiddleToken() {
        // prefix=1 (skip "hello"), midEnd=2 (stop before "world")
        String[] tokens = new String[]{"hello", " ", "world"};
        int[] span = WordDiff.changedRange(tokens, 1, 2);
        // start = length("hello") = 5; end = 5 + length(" ") = 6
        assertArrayEquals(new int[]{5, 6}, span);
    }

    @Test
    void changedRangeEmptySpan() {
        String[] tokens = new String[]{"hello", " ", "world"};
        int[] span = WordDiff.changedRange(tokens, 2, 2);
        // start = length("hello") + length(" ") = 6; end = 6
        assertArrayEquals(new int[]{6, 6}, span);
    }

    // ------------------------------------------------------------------ //
    //  compute() — main algorithm                                          //
    // ------------------------------------------------------------------ //

    @Test
    void computeIdenticalLinesReturnsNull() {
        assertNull(WordDiff.compute("hello world", "hello world"),
                "token-identical lines should return null (no highlight)");
    }

    @Test
    void computeSingleWordDiff() {
        // "status=active" vs "status=inactive"
        // Prefix: "status", "=" — 2 tokens shared.
        // Suffix: nothing (after common prefix, "active" vs "inactive" differ entirely).
        var result = WordDiff.compute("status=active", "status=inactive");
        assertNotNull(result);
        // Left span covers "active" starting after "status="
        assertEquals(7, result.left().start());
        assertEquals(13, result.left().end());   // "active" is 6 chars → 7+6=13
        // Right span covers "inactive" starting after "status="
        assertEquals(7, result.right().start());
        assertEquals(15, result.right().end());  // "inactive" is 8 chars → 7+8=15
    }

    @Test
    void computeLinesTooLongReturnsNull() {
        String longString = "a".repeat(15_000);
        assertNull(WordDiff.compute(longString, longString + "b"),
                "lines exceeding INLINE_MAX_CHARS combined should return null");
    }

    @Test
    void computeEmptyVsNonEmpty() {
        var result = WordDiff.compute("", "hello");
        assertNotNull(result);
        // Left span is empty [0,0], right span covers all of "hello" [0,5]
        assertEquals(0, result.left().start());
        assertEquals(0, result.left().end());
        assertEquals(0, result.right().start());
        assertEquals(5, result.right().end());
    }

    @Test
    void computeNonEmptyVsEmpty() {
        var result = WordDiff.compute("hello", "");
        assertNotNull(result);
        assertEquals(0, result.left().start());
        assertEquals(5, result.left().end());
        assertEquals(0, result.right().start());
        assertEquals(0, result.right().end());
    }

    @Test
    void computeSpansAreCorrectCharOffsets() {
        // "id=1000 name=alice status=active" vs "id=1000 name=alice status=inactive"
        // Common prefix tokens: id, =, 1000, ' ', name, =, alice, ' ', status, =
        // That prefix text = "id=1000 name=alice status="  (length = 26)
        // Changed token: left="active" right="inactive"
        var result = WordDiff.compute(
                "id=1000 name=alice status=active",
                "id=1000 name=alice status=inactive");
        assertNotNull(result);
        assertEquals(26, result.left().start());
        assertEquals(32, result.left().end());   // "active" = 6 chars
        assertEquals(26, result.right().start());
        assertEquals(34, result.right().end());  // "inactive" = 8 chars
    }

    @Test
    void computeChangedMiddleWithCommonSuffix() {
        // left:  "BEGIN foo END"
        // right: "BEGIN foobar END"
        // Prefix: "BEGIN", " "  -> 2 tokens
        // Suffix: " ", "END"   -> 2 tokens (suffix can only use maxSuffix = min(3,4)-2 = 1 here)
        // Actually: a=["BEGIN"," ","foo"," ","END"], b=["BEGIN"," ","foobar"," ","END"]
        // prefix=2 ("BEGIN"," "), maxSuffix=min(5,5)-2=3
        // suffix check: a[4]="END"==b[4]="END" -> suffix=1; a[3]=" "==b[3]=" " -> suffix=2;
        //               a[2]="foo" != b[2]="foobar" -> stop at suffix=2
        // left midStart=2, midEnd=5-2=3 -> token "foo" -> charRange = 6..9
        // right midStart=2, midEnd=5-2=3 -> token "foobar" -> charRange = 6..12
        var result = WordDiff.compute("BEGIN foo END", "BEGIN foobar END");
        assertNotNull(result);
        assertEquals(6, result.left().start());
        assertEquals(9, result.left().end());
        assertEquals(6, result.right().start());
        assertEquals(12, result.right().end());
    }

    @Test
    void computeInlineMaxCharsThreshold() {
        // Exactly at the threshold: should NOT be null
        int halfMax = WordDiff.INLINE_MAX_CHARS / 2;
        String left  = "a".repeat(halfMax);
        String right = "b".repeat(halfMax);
        // combined = 20_000 exactly — the guard is "> 20_000" (strictly greater), so this is OK
        assertNotNull(WordDiff.compute(left, right));

        // One over the threshold: should be null
        assertNull(WordDiff.compute(left + "x", right));
    }
}

File: jetbrains/src/test/kotlin/engine/WordDiffTest.kt
package com.adityakumar.engine

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class WordDiffTest {

    // ── tokenize ───────────────────────────────────────────────────────────────

    @Test fun `tokenize empty string returns empty list`() {
        assertEquals(emptyList<String>(), WordDiff.tokenize(""))
    }

    @Test fun `tokenize pure word`() {
        assertEquals(listOf("hello"), WordDiff.tokenize("hello"))
    }

    @Test fun `tokenize word with trailing space`() {
        // "hello " → ["hello", " "]
        assertEquals(listOf("hello", " "), WordDiff.tokenize("hello "))
    }

    @Test fun `tokenize mixed content`() {
        // "foo, bar" → ["foo", ",", " ", "bar"]
        assertEquals(listOf("foo", ",", " ", "bar"), WordDiff.tokenize("foo, bar"))
    }

    @Test fun `tokenize whitespace run is one token`() {
        assertEquals(listOf("a", "   ", "b"), WordDiff.tokenize("a   b"))
    }

    @Test fun `tokenize punctuation run is one token`() {
        assertEquals(listOf("abc", "...", "def"), WordDiff.tokenize("abc...def"))
    }

    @Test fun `tokenize concatenates back to original`() {
        val text = "SELECT id, name FROM users WHERE id = 42;"
        val tokens = WordDiff.tokenize(text)
        assertEquals(text, tokens.joinToString(""))
    }

    @Test fun `tokenize digits are word chars`() {
        assertEquals(listOf("123"), WordDiff.tokenize("123"))
    }

    // ── changedRange ───────────────────────────────────────────────────────────

    @Test fun `changedRange empty token list`() {
        val span = WordDiff.changedRange(emptyList(), 0, 0)
        assertEquals(WordDiff.Span(0, 0), span)
    }

    @Test fun `changedRange full span`() {
        val tokens = listOf("foo", " ", "bar")
        // midStart=0, midEnd=3 → start=0, end=7
        val span = WordDiff.changedRange(tokens, 0, 3)
        assertEquals(WordDiff.Span(0, 7), span)
    }

    @Test fun `changedRange middle span`() {
        val tokens = listOf("a", "b", "c", "d")  // lengths 1,1,1,1
        // midStart=1, midEnd=3 → start=1, end=3
        val span = WordDiff.changedRange(tokens, 1, 3)
        assertEquals(WordDiff.Span(1, 3), span)
    }

    @Test fun `changedRange when midStart equals midEnd returns empty span`() {
        val tokens = listOf("hello", " ", "world")
        val span = WordDiff.changedRange(tokens, 2, 2)
        // start = len("hello") + len(" ") = 6; end = 6
        assertEquals(WordDiff.Span(6, 6), span)
    }

    // ── computeInline ──────────────────────────────────────────────────────────

    @Test fun `computeInline identical lines returns null (no highlight)`() {
        assertNull(WordDiff.computeInline("hello world", "hello world"))
    }

    @Test fun `computeInline empty strings returns null`() {
        assertNull(WordDiff.computeInline("", ""))
    }

    @Test fun `computeInline one empty one non-empty`() {
        val result = WordDiff.computeInline("", "hello")
        assertNotNull(result)
        // left span is empty (nothing in left), right span covers "hello"
        assertEquals(WordDiff.Span(0, 0), result!!.left)
        assertEquals(WordDiff.Span(0, 5), result.right)
    }

    @Test fun `computeInline completely different words`() {
        val result = WordDiff.computeInline("foo", "bar")
        assertNotNull(result)
        assertEquals(WordDiff.Span(0, 3), result!!.left)
        assertEquals(WordDiff.Span(0, 3), result.right)
    }

    @Test fun `computeInline common prefix trimmed`() {
        // "status=active" vs "status=inactive" — common prefix "status="
        val result = WordDiff.computeInline("status=active", "status=inactive")
        assertNotNull(result)
        // tokens left:  ["status", "=", "active"]
        // tokens right: ["status", "=", "inactive"]
        // common prefix = 2 tokens ("status", "="), no common suffix
        // left span: offset 7 to 13 ("active")
        // right span: offset 7 to 15 ("inactive")
        assertEquals(WordDiff.Span(7, 13), result!!.left)
        assertEquals(WordDiff.Span(7, 15), result.right)
    }

    @Test fun `computeInline common suffix trimmed`() {
        // "hello world" vs "goodbye world" — common suffix " world"
        val result = WordDiff.computeInline("hello world", "goodbye world")
        assertNotNull(result)
        // tokens: ["hello", " ", "world"] vs ["goodbye", " ", "world"]
        // prefix=0, suffix=2 (" ", "world")
        // left span: 0..5 ("hello")
        // right span: 0..7 ("goodbye")
        assertEquals(WordDiff.Span(0, 5), result!!.left)
        assertEquals(WordDiff.Span(0, 7), result.right)
    }

    @Test fun `computeInline common prefix AND suffix`() {
        // "id=1000 name=alice status=active" vs "id=1000 name=alice status=inactive"
        val left  = "id=1000 name=alice status=active"
        val right = "id=1000 name=alice status=inactive"
        val result = WordDiff.computeInline(left, right)
        assertNotNull(result)
        // Both spans should be non-empty
        assertTrue(result!!.left.end > result.left.start)
        assertTrue(result.right.end > result.right.start)
        // The changed part is only at the end ("active" vs "inactive")
        // Left span ends at left.length, right span ends at right.length
        assertEquals(left.length, result.left.end)
        assertEquals(right.length, result.right.end)
    }

    @Test fun `computeInline returns null when total chars exceed 20000`() {
        val big = "a".repeat(10_001)
        assertNull(WordDiff.computeInline(big, big.replace("a", "b")))
    }

    @Test fun `computeInline boundary exactly at limit returns result`() {
        // 10000 + 10000 = 20000 — exactly at limit, not over, should compute
        val left  = "a".repeat(10_000)
        val right = "b".repeat(10_000)
        // If sum == 20000 we allow it; sum > 20000 we reject
        val result = WordDiff.computeInline(left, right)
        assertNotNull(result)
    }

    @Test fun `computeInline single character change`() {
        val result = WordDiff.computeInline("abc", "axc")
        assertNotNull(result)
        // tokens: ["abc"] vs ["axc"] — no common prefix, no common suffix at token level
        // Entire strings are highlighted since single tokens differ
        assertEquals(WordDiff.Span(0, 3), result!!.left)
        assertEquals(WordDiff.Span(0, 3), result.right)
    }

    @Test fun `computeInline whitespace-only difference`() {
        // "a b" vs "a  b" — the space token differs (one space vs two spaces)
        val result = WordDiff.computeInline("a b", "a  b")
        assertNotNull(result)
        // tokens: ["a", " ", "b"] vs ["a", "  ", "b"]
        // prefix=1 ("a"), suffix=1 ("b"), middle = [" "] vs ["  "]
        assertEquals(WordDiff.Span(1, 2), result!!.left)
        assertEquals(WordDiff.Span(1, 3), result.right)
    }
}
